README.md
# videojs-upnext

Youtube like plugin for up next videos

## Installation

```sh
npm install --save videojs-upnext
```

## Usage

To include videojs-upnext on your website or web application, use any of the following methods.

### `<script>` Tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available.

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/videojs-upnext.min.js"></script>
<script>
  var player = videojs('my-video');
  player.upnext({
    timeout : 5000,
    headText : 'Up Next',
    cancelText: 'Cancel',
    getTitle : function() { return 'Next video title' },
    next : function () { performActionAfterTimeout() }
  });
</script>
```

### Browserify

When using with Browserify, install videojs-upnext via npm and `require` the plugin as you would any other module.

```js
var videojs = require('video.js');

// The actual plugin function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require('videojs-upnext');

var player = videojs('my-video');

player.upnext({
  timeout : 5000,
  headText : 'Up Next',
  cancelText: 'Cancel',
  getTitle : function() { return 'Next video title' },
  next : function () { performActionAfterTimeout() }
});
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(['video.js', 'videojs-upnext'], function(videojs) {
  var player = videojs('my-video');

  player.upnext({
    timeout : 5000,
    headText : 'Up Next',
    cancelText: 'Cancel',
    getTitle : function() { return 'Next video title' },
    next : function () { performActionAfterTimeout() }
  });
});
```

## License

Apache-2.0. Copyright (c) Fernando Godino <fernando@varsityviews.com>


[videojs]: http://videojs.com/


CHANGELOG.md
CHANGELOG
=========

## HEAD (Unreleased)
_(none)_

--------------------

## 2.2.0 (2017-05-09)
_(none)_



test/index.test.js
import document from 'global/document';
import QUnit from 'qunitjs';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/js/index.js';

const Player = videojs.getComponent('Player');

QUnit.module('sanity tests');

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-upnext', {

  beforeEach() {

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5. This MUST come
    // before any player is created; otherwise, timers could get created
    // with the actual timer methods!
    this.clock = sinon.useFakeTimers();

    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);
  },

  afterEach() {
    this.player.dispose();
    this.clock.restore();
  }
});

QUnit.test('registers itself with video.js', function(assert) {
  assert.expect(2);

  assert.strictEqual(
    typeof Player.prototype.upnext,
    'function',
    'videojs-upnext plugin was registered'
  );

  this.player.upnext();

  // Tick the clock forward enough to trigger the player to be "ready".
  this.clock.tick(1);

  assert.ok(
    this.player.hasClass('vjs-upnext'),
    'the plugin adds a class to the player'
  );
});


test/karma.conf.js

module.exports = function(config) {
  var detectBrowsers = {
    enabled: false,
    usePhantomJS: false
  };

  // On Travis CI, we can only run in Firefox.
  if (process.env.TRAVIS) {
    config.browsers = ['Firefox', 'travisChrome'];
  }

  // If no browsers are specified, we enable `karma-detect-browsers`
  // this will detect all browsers that are available for testing
  if (!config.browsers.length) {
    detectBrowsers.enabled = true;
  }

  config.set({
    basePath: '..',
    frameworks: ['browserify', 'qunit', 'detectBrowsers'],

    files: [
      'dist/videojs-upnext.css',
      'node_modules/sinon/pkg/sinon.js',
      'node_modules/sinon/pkg/sinon-ie.js',
      'node_modules/video.js/dist/video.js',
      'test/**/*.js'
    ],
    exclude: [
      'test/bundle.js'
    ],

    customLaunchers: {
      travisChrome: {
        base: 'Chrome',
        flags: ['--no-sandbox']
      }
    },

    preprocessors: {
      'test/**/*.js': ['browserify']
    },

    detectBrowsers: detectBrowsers,
    reporters: ['dots'],
    port: 9876,
    colors: true,
    autoWatch: false,
    singleRun: true,
    concurrency: Infinity,

    browserify: {
      transform: [
        'babelify',
        'browserify-shim'
      ]
    }
  });
};


test/plugin.test.js
import document from 'global/document';

import QUnit from 'qunit';
import sinon from 'sinon';
import videojs from 'video.js';

import plugin from '../src/plugin';

const Player = videojs.getComponent('Player');

QUnit.test('the environment is sane', function(assert) {
  assert.strictEqual(typeof Array.isArray, 'function', 'es5 exists');
  assert.strictEqual(typeof sinon, 'object', 'sinon exists');
  assert.strictEqual(typeof videojs, 'function', 'videojs exists');
  assert.strictEqual(typeof plugin, 'function', 'plugin is a function');
});

QUnit.module('videojs-upnext', {

  beforeEach() {
    this.fixture = document.getElementById('qunit-fixture');
    this.video = document.createElement('video');
    this.fixture.appendChild(this.video);
    this.player = videojs(this.video);

    // Mock the environment's timers because certain things - particularly
    // player readiness - are asynchronous in video.js 5.
    this.clock = sinon.useFakeTimers();
  },

  afterEach() {

    // The clock _must_ be restored before disposing the player; otherwise,
    // certain timeout listeners that happen inside video.js may throw errors.
    this.clock.restore();
    this.player.dispose();
  }
});

QUnit.test('registers itself with video.js', function(assert) {

  assert.strictEqual(
    typeof Player.prototype.upnext,
    'function',
    'videojs-upnext plugin was registered'
  );

  this.player.upnext();

  // Tick the clock forward enough to trigger the player to be "ready".
  this.clock.tick(1);
});


src/plugin.js
import videojs from 'video.js';

const registerPlugin = videojs.registerPlugin || videojs.plugin;
let Component = videojs.getComponent('Component');


function getMainTemplate(options) {
  return `
    <div class="vjs-upnext-top">
      <span class="vjs-upnext-headtext">${options.headText}</span>
      <div class="vjs-upnext-title"></div>
    </div>
    <div class="vjs-upnext-autoplay-icon">
      <svg height="100%" version="1.1" viewbox="0 0 98 98" width="100%">
        <circle class="vjs-upnext-svg-autoplay-circle" cx="49" cy="49" fill="#000" fill-opacity="0.8" r="48"></circle>
        <circle class="vjs-upnext-svg-autoplay-ring" cx="-49" cy="49" fill-opacity="0" r="46.5" stroke="#FFFFFF" stroke-width="4" transform="rotate(-90)"></circle>
        <polygon class="vjs-upnext-svg-autoplay-triangle" fill="#fff" points="32,27 72,49 32,71"></polygon></svg>
    </div>
    <span class="vjs-upnext-bottom">
      <span class="vjs-upnext-cancel">
        <button class="vjs-upnext-cancel-button" tabindex="0" aria-label="Cancel autoplay">${options.cancelText}</button>
      </span>
    </span>
  `;
}

/**
 * EndCard Component
 */
export class EndCard extends Component {

  constructor(player, options) {
    super(player, options);

    this.getTitle = this.options_.getTitle;
    this.next = this.options_.next;

    this.upNextEvents = new videojs.EventTarget();

    this.dashOffsetTotal = 586;
    this.dashOffsetStart = 293;
    this.interval = 50;
    this.chunkSize = (this.dashOffsetTotal - this.dashOffsetStart) / (this.options_.timeout / this.interval);

    player.on('ended', (event) => {
      player.addClass('vjs-upnext');
      this.showCard((canceled) => {
        player.removeClass('vjs-upnext');
        this.container.style.display = 'none';
        if (!canceled) {
          this.next();
        }
      });
    });

    player.on('playing', function() {
      this.upNextEvents.trigger('playing');
    }.bind(this));
  }

  createEl() {

    let container = super.createEl('div', {
      className: 'vjs-upnext-content',
      innerHTML: getMainTemplate(this.options_)
    });

    this.container = container;
    container.style.display = 'none';

    this.autoplayRing = container.getElementsByClassName('vjs-upnext-svg-autoplay-ring')[0];
    this.title = container.getElementsByClassName('vjs-upnext-title')[0];
    this.cancelButton = container.getElementsByClassName('vjs-upnext-cancel-button')[0];
    this.nextButton = container.getElementsByClassName('vjs-upnext-autoplay-icon')[0];

    this.cancelButton.onclick = function() {
      this.upNextEvents.trigger('cancel');
    }.bind(this);

    this.nextButton.onclick = function() {
      this.upNextEvents.trigger('next');
    }.bind(this);

    return container;
  }

  showCard(cb) {

    let timeout;
    let start;
    let now;
    let newOffset;

    this.autoplayRing.setAttribute('stroke-dasharray', this.dashOffsetStart);
    this.autoplayRing.setAttribute('stroke-dashoffset', -this.dashOffsetStart);

    this.title.innerHTML = this.getTitle();

    this.upNextEvents.one('cancel', () => {
      clearTimeout(timeout);
      cb(true);
    });

    this.upNextEvents.one('playing', () => {
      clearTimeout(timeout);
      cb(true);
    });

    this.upNextEvents.one('next', () => {
      clearTimeout(timeout);
      cb(false);
    });

    let update = function() {
      now = this.options_.timeout - (new Date().getTime() - start);

      if (now <= 0) {
        clearTimeout(timeout);
        cb(false);
      } else {
        newOffset = Math.max(-this.dashOffsetTotal, this.autoplayRing.getAttribute('stroke-dashoffset') - this.chunkSize);
        this.autoplayRing.setAttribute('stroke-dashoffset', newOffset);
        timeout = setTimeout(update.bind(this), this.interval);
      }

    };

    this.container.style.display = 'block';
    start = new Date().getTime();
    timeout = setTimeout(update.bind(this), this.interval);
  }
}

videojs.registerComponent('EndCard', EndCard);

/**
 * A video.js plugin.
 *
 * In the plugin function, the value of `this` is a video.js `Player`
 * instance. You cannot rely on the player being in a "ready" state here,
 * depending on how the plugin is invoked. This may or may not be important
 * to you; if not, remove the wait for "ready"!
 *
 * @function upnext
 * @param    {Object} [options={}]
 *           An object of options left to the plugin author to define.
 */
const upnext = function(options) {
  let opts = options || {};
  let settings = {
    next: opts.next,
    getTitle: opts.getTitle,
    timeout: opts.timeout || 5000,
    cancelText: opts.cancelText || 'Cancel',
    headText: opts.headText || 'Up Next'
  };

  this.addChild('endCard', settings);

};

// Register the plugin with video.js.
registerPlugin('upnext', upnext);

export default upnext;


src/js/index.js
import videojs from 'video.js';

const defaults = {};

// Cross-compatibility for Video.js 5 and 6.
const registerPlugin = videojs.registerPlugin || videojs.plugin;
// const dom = videojs.dom || videojs;

function getMainTemplate(options) {
  return `
    <div class="vjs-upnext-top">
      <span class="vjs-upnext-headtext">${options.headText}</span>
      <div class="vjs-upnext-title"></div>
    </div>
    <div class="vjs-upnext-autoplay-icon">
      <svg height="100%" version="1.1" viewbox="0 0 98 98" width="100%">
        <circle class="vjs-upnext-svg-autoplay-circle" cx="49" cy="49" fill="#000" fill-opacity="0.8" r="48"></circle>
        <circle class="vjs-upnext-svg-autoplay-ring" cx="-49" cy="49" fill-opacity="0" r="46.5" stroke="#FFFFFF" stroke-width="4" transform="rotate(-90)"></circle>
        <polygon class="vjs-upnext-svg-autoplay-triangle" fill="#fff" points="32,27 72,49 32,71"></polygon></svg>
    </div>
    <span class="vjs-upnext-bottom">
      <span class="vjs-upnext-cancel">
        <button class="vjs-upnext-cancel-button" tabindex="0" aria-label="Cancel autoplay">${options.cancelText}</button>
      </span>
    </span>
  `;
}

const Component = videojs.getComponent('Component');

/**
 * EndCard Component
 */
export class EndCard extends Component {

  constructor(player, options) {
    super(player, options);

    this.getTitle = this.options_.getTitle;
    this.next = this.options_.next;

    this.upNextEvents = new videojs.EventTarget();

    this.dashOffsetTotal = 586;
    this.dashOffsetStart = 293;
    this.interval = 50;
    this.chunkSize = (this.dashOffsetTotal - this.dashOffsetStart) / (this.options_.timeout / this.interval);

    player.on('ended', (event) => {
      player.addClass('vjs-upnext--showing');
      this.showCard((canceled) => {
        player.removeClass('vjs-upnext--showing');
        this.container.style.display = 'none';
        if (!canceled) {
          this.next();
        }
      });
    });

    player.on('playing', function() {
      this.upNextEvents.trigger('playing');
    }.bind(this));
  }

  createEl() {

    const container = super.createEl('div', {
      className: 'vjs-upnext-content',
      innerHTML: getMainTemplate(this.options_)
    });

    this.container = container;
    container.style.display = 'none';

    this.autoplayRing = container.getElementsByClassName('vjs-upnext-svg-autoplay-ring')[0];
    this.title = container.getElementsByClassName('vjs-upnext-title')[0];
    this.cancelButton = container.getElementsByClassName('vjs-upnext-cancel-button')[0];
    this.nextButton = container.getElementsByClassName('vjs-upnext-autoplay-icon')[0];

    this.cancelButton.onclick = function() {
      this.upNextEvents.trigger('cancel');
    }.bind(this);

    this.nextButton.onclick = function() {
      this.upNextEvents.trigger('next');
    }.bind(this);

    return container;
  }

  showCard(cb) {

    let timeout;
    let start;
    let now;
    let newOffset;

    this.autoplayRing.setAttribute('stroke-dasharray', this.dashOffsetStart);
    this.autoplayRing.setAttribute('stroke-dashoffset', -this.dashOffsetStart);

    this.title.innerHTML = this.getTitle();

    this.upNextEvents.one('cancel', () => {
      clearTimeout(timeout);
      cb(true);
    });

    this.upNextEvents.one('playing', () => {
      clearTimeout(timeout);
      cb(true);
    });

    this.upNextEvents.one('next', () => {
      clearTimeout(timeout);
      cb(false);
    });

    const update = function() {
      now = this.options_.timeout - (new Date().getTime() - start);

      if (now <= 0) {
        clearTimeout(timeout);
        cb(false);
      } else {
        newOffset = Math.max(-this.dashOffsetTotal, this.autoplayRing.getAttribute('stroke-dashoffset') - this.chunkSize);
        this.autoplayRing.setAttribute('stroke-dashoffset', newOffset);
        timeout = setTimeout(update.bind(this), this.interval);
      }

    };

    this.container.style.display = 'block';
    start = new Date().getTime();
    timeout = setTimeout(update.bind(this), this.interval);
  }
}

videojs.registerComponent('EndCard', EndCard);

/**
 * Function to invoke when the player is ready.
 *
 * This is a great place for your plugin to initialize itself. When this
 * function is called, the player will have its DOM and child components
 * in place.
 *
 * @function onPlayerReady
 * @param    {Player} player
 *           A Video.js player.
 * @param    {Object} [options={}]
 *           An object of options left to the plugin author to define.
 */
const onPlayerReady = (player, options) => {
  player.addClass('vjs-upnext');
};

/**
 * A video.js plugin.
 *
 * In the plugin function, the value of `this` is a video.js `Player`
 * instance. You cannot rely on the player being in a "ready" state here,
 * depending on how the plugin is invoked. This may or may not be important
 * to you; if not, remove the wait for "ready"!
 *
 * @function upnext
 * @param    {Object} [options={}]
 *           An object of options left to the plugin author to define.
 */
const upnext = function(options) {

  this.ready(() => {
    onPlayerReady(this, videojs.mergeOptions(defaults, options));
  });

  const opts = options || {};
  const settings = {
    next: opts.next,
    getTitle: opts.getTitle,
    timeout: opts.timeout || 5000,
    cancelText: opts.cancelText || 'Cancel',
    headText: opts.headText || 'Up Next'
  };

  this.addChild('endCard', settings);

};

// Register the plugin with video.js.
registerPlugin('upnext', upnext);

// Include the version number.
upnext.VERSION = '__VERSION__';

export default upnext;
